go/engine/identify2_with_uid.go

```diff
@@ -0,0 +1,457 @@
+// we can unblock the caller.
+func (e *Identify2WithUID) CCLCheckCompleted(lcr *libkb.LinkCheckResult) {
+	e.remotesMutex.Lock()
+	defer e.remotesMutex.Unlock()
+
+	e.G().Log.Debug("+ CheckCompleted for %s", lcr.GetLink().ToIDString())
+	defer e.G().Log.Debug("- CheckCompleted")
+
+	if !e.useRemoteAssertions() || e.useTracking {
+		e.G().Log.Debug("| Early out, since not using remote assertions or is tracking")
+		return
+	}
+
+	pf := libkb.RemoteProofChainLinkToProof(lcr.GetLink())
+	e.remotesReceived.Add(pf)
+
+	if !e.remoteAssertion.HasFactor(pf) {
+		e.G().Log.Debug("| Proof isn't needed in our remote-assertion early-out check: %v", pf)
+		return
+	}
+
+	if err := lcr.GetError(); err != nil {
+		e.G().Log.Debug("| got error -> %v", err)
+		e.remotesError = err
+	}
+
+	// note(maxtaco): this is a little ugly in that it's O(n^2) where n is the number
+	// of identities in the assertion. But I can't imagine n > 3, so this is fine
+	// for now.
+	matched := e.remoteAssertion.MatchSet(*e.remotesReceived)
+	e.G().Log.Debug("| matched -> %v", matched)
+	if matched {
+		e.remotesCompleted = true
+	}
+
+	if e.remotesError != nil || e.remotesCompleted {
+		e.G().Log.Debug("| unblocking, with err = %v", e.remotesError)
+		e.unblock(e.remotesError)
+	}
+}
+
+func (e *Identify2WithUID) checkLocalAssertions() error {
+	if !e.localAssertion.MatchSet(*e.them.BaseProofSet()) {
+		return libkb.UnmetAssertionError{User: e.them.GetName(), Remote: false}
+	}
+	return nil
+}
+
+func (e *Identify2WithUID) checkRemoteAssertions(okStates []keybase1.ProofState) error {
+	ps := libkb.NewProofSet(nil)
+	e.state.Result().AddProofsToSet(ps, okStates)
+	if !e.remoteAssertion.MatchSet(*ps) {
+		return libkb.UnmetAssertionError{User: e.them.GetName(), Remote: true}
+	}
+	return nil
+}
+
+func (e *Identify2WithUID) loadAssertion() (err error) {
+	e.themAssertion, err = libkb.AssertionParseAndOnly(e.arg.UserAssertion)
+	if err == nil {
+		e.remoteAssertion, e.localAssertion = libkb.CollectAssertions(e.themAssertion)
+	}
+	return err
+}
+
+func (e *Identify2WithUID) useAnyAssertions() bool {
+	return e.useLocalAssertions() || e.useRemoteAssertions()
+}
+
+func (e *Identify2WithUID) useLocalAssertions() bool {
+	return e.localAssertion.Len() > 0
+}
+func (e *Identify2WithUID) useRemoteAssertions() bool {
+	return e.remoteAssertion.Len() > 0
+}
+
+func (e *Identify2WithUID) runIdentifyPrecomputation() (err error) {
+	f := func(k keybase1.IdentifyKey) {
+		e.identifyKeys = append(e.identifyKeys, k)
+	}
+	e.state.Precompute(f)
+	return nil
+}
+
+func (e *Identify2WithUID) runIdentifyUI(ctx *Context) (err error) {
+	e.G().Log.Debug("+ runIdentifyUI(%s)", e.them.GetName())
+
+	ctx.IdentifyUI.Start(e.them.GetName())
+	for _, k := range e.identifyKeys {
+		ctx.IdentifyUI.DisplayKey(k)
+	}
+	ctx.IdentifyUI.ReportLastTrack(libkb.ExportTrackSummary(e.state.TrackLookup(), e.them.GetName()))
+	ctx.IdentifyUI.LaunchNetworkChecks(e.state.ExportToUncheckedIdentity(), e.them.Export())
```

Comment:
Do Start, DisplayKey, ReportLastTrack, and LaunchNetworkChecks always happen at the same time?  If so, maybe protocol should change to consolidate this?